### ISSUE TITLE
Improved validation for IIS steps is causing unexpected behavior

### ISSUE THREAD

**Issue by danielfischeroctopus** [OctopusDeploy/Issues#3484]
The following change https://github.com/OctopusDeploy/Issues/issues/3411 has caused some unexpected behavior with IIS steps. The following are two issues observed so far.

First:
If you use a custom expression with one of the prohibited values, you are unable to save/edit the step.
E.g. - `#{URL:WebSite}` Will fail to save due to the `:` even if it will evaluate to an allowed value.
Source: https://secure.helpscout.net/conversation/361787125/15528/?folderId=845288

Second:
**Update**
In this case more than likely it is a different step that has incorrect data. E.g. the user tries to update step 4 but it is step 1 that has incorrect data.

For pre-existing projects that were created before 3.13.1, you can not edit/save/create any IIS steps. They throw the following error regardless of what you enter as the value. (This error does not display if you add the step to a project that was created post 3.13.1)

```
There was a problem with your request.
The IIS web site name cannot contain any of these characters \, /, ?, ;, :, @, &, =, +, $, ,, |, ", <, >
IIS application pool name cannot contain any of these characters \, /, ", |, <, >, :, *, ?, ], [, +, =, ;, ,
```

Source: https://secure.helpscout.net/conversation/362745269/15583/?folderId=845288

Attached is a screenshot from my server where I replicated this. Steps to replicate:
Create a new IIS step (I used `IIS Application - Create`) on a project that existed before 3.13.1
Attempt to save step with normally allowed values.

![template](https://cloud.githubusercontent.com/assets/18252832/25926908/855f48c4-3635-11e7-91f8-2d6e2274c2ef.JPG)

**Comment by danielfischeroctopus** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Improved validation for IIS steps is causing unexpected behavior](https://trello.com/c/lzA17Hx8/545-improved-validation-for-iis-steps-is-causing-unexpected-behavior)

**Comment by pawelpabich** [comment]
Release Note: Variables with colons can now be used as Web Site and App Pool names

**Comment by timothybeutels** [comment]
Since the release note explicitly mentions colons; I wanted to point out that the validation also breaks when using the substitution syntax to reference an output variable of another step.

For example:
#{Octopus.Action[Modify Web Configuration Variables].Output.WebsiteName}

**Comment by creekg** [comment]
I am using 3.13.6 and this issue is NOT fixed - I am unable to make a very small but important change to  a complex deployment step in a large complex project. There is no way I can recreate this whole project from scratch before my deployment deadline.

**Comment by pawelpabich** [comment]
Hi @creekg,

It looks like we've missed an edge case. Sorry about that. To get to the bottom your problem as quickly as possible I would need you to provide me with two pieces of information. Once you have them please send them to us using our [support email address](https://octopus.com/support). Please don't post them here as more than likely they will contain sensitive information.

- A screenshot of the error
- A JSON snapshot of your deployment process. You can grab it directly using a web browser. All you have to do is to login to Octopus and then navigate to  `http://youroctopusserver/api/deploymentprocesses/deploymentprocess-Projects-X` where `Projects-X` is the Id of your project, e.g. `Projects-1`

**Comment by kennethhovsnow** [comment]
Hi! I'm using 3.16.7 and experience the same error with IIS poolname. I use a pre-deploy script to set poolname. My poolname is #{Octopus.Action[Deploy new API].Output.Poolname}
The same error occurs when I'm trying to clone the step.

**Comment by pawelpabich** [comment]
Hi @kennethhovsnow,

It looks like another edge case. Can you please follow the instructions above your comment so we can get to the bottom of your problem?

**Comment by kennethhovsnow** [comment]
Hmm I'm not able to get the JSON with the api url provided above. I just get error: 
{
"ErrorMessage": "The resource 'deploymentprocess-Projects-myproject' was not found." Myproject is the name of my project.
}

**Comment by kennethhovsnow** [comment]
I had to goto http://youroctopusserver/api/deploy/api/projects/all to find the project id

**Comment by pawelpabich** [comment]
Great, I'm glad you've found it.

**Comment by kennethhovsnow** [comment]
This should be a high importance bug fix, because after upgrade to 3.16.7 I am not able to make changes to the step using output variable in the poolname, because square brackets are not allowed

**Comment by kennethhovsnow** [comment]
We just upgraded to 3.17.5, and I got the same issue. Not able to save or clone the IIS step because poolname contains [ ]. Then I see that the step it is complaining about is not the same step I am trying to save (“Deploy API” get error, but I trying to save “Deploy new API”). Then I go to Deploy API, and click save. No errors. Then I go back to Deploy new API and everything is fine. I am now able to edit and clone the step.
